kong/dns/client.lua

```diff
@@ -0,0 +1,824 @@
+  return cache:get("last:" .. name)
+end
+
+
+local init_hosts do
+  local function insert_answer_into_cache(cache, hosts_cache, address, name, qtype)
+    local key = name .. ":" .. qtype
+    local answers = {
+      ttl = LONG_LASTING_TTL,
+      expire = now() + LONG_LASTING_TTL,
+      {
+        name = name,
+        type = qtype,
+        address = address,
+        class = 1,
+        ttl = LONG_LASTING_TTL,
+      },
+    }
+
+    -- insert via the `:get` callback to prevent inter-process communication
+    cache:get(key, nil, function()
+      return answers, nil, LONG_LASTING_TTL
+    end)
+
+    -- used for the host entry eviction
+    hosts_cache[key] = answers
+  end
+
+  -- insert hosts into cache
+  function init_hosts(cache, path, preferred_ip_type)
+    local hosts = parse_hosts(path)
+    local hosts_cache = {}
+
+    for name, address in pairs(hosts) do
+      name = string_lower(name)
+
+      if address.ipv4 then
+        insert_answer_into_cache(cache, hosts_cache, address.ipv4, name, TYPE_A)
+        insert_last_type(cache, name, TYPE_A)
+      end
+
+      if address.ipv6 then
+        insert_answer_into_cache(cache, hosts_cache, address.ipv6, name, TYPE_AAAA)
+        if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+          insert_last_type(cache, name, TYPE_AAAA)
+        end
+      end
+    end
+
+    return hosts, hosts_cache
+  end
+end
+
+
+-- distinguish the worker_events sources registered by different new() instances
+local ipc_counter = 0
+
+function _M.new(opts)
+  opts = opts or {}
+
+  -- parse resolv.conf
+  local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, opts.enable_ipv6)
+  if not resolv then
+    log(WARN, "Invalid resolv.conf: ", err)
+    resolv = { options = {} }
+  end
+
+  -- init the resolver options for lua-resty-dns
+  local nameservers = (opts.nameservers and not table_isempty(opts.nameservers))
+                      and opts.nameservers
+                      or resolv.nameservers
+
+  if not nameservers or table_isempty(nameservers) then
+    log(WARN, "Invalid configuration, no nameservers specified")
+  end
+
+  local r_opts = {
+    retrans = opts.retrans or resolv.options.attempts or 5,
+    timeout = opts.timeout or resolv.options.timeout or 2000, -- ms
+    no_random = opts.no_random or not resolv.options.rotate,
+    nameservers = nameservers,
+  }
+
+  -- init the mlcache
+
+  -- maximum timeout for the underlying r:query() operation to complete
+  -- socket timeout * retrans * 2 calls for send and receive + 1s extra delay
+  local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans * 2 + 1 -- s
+
+  local resty_lock_opts = {
+    timeout = lock_timeout,
+    exptimeout = lock_timeout + 1,
+  }
+
+  -- TODO: convert the ipc a module constant, currently we need to use the
+  --       ipc_source to distinguish sources of different DNS client events.
+  ipc_counter = ipc_counter + 1
+  local ipc_source = "dns_client_mlcache#" .. ipc_counter
```

Comment:
I don't know how beneficial would it be to optimize it, but what if instead of a counter we used the hash of `opts`, we would use better the cache, wouldn't we?

Reply:
wow! that's what i want to do here, let me review code and test whether the multiple (re)configurations of dns client will generate different hash of opts. If the hash value could be unique every time of reconfiguration. I'll accept your advice.

Reply:
After some work, I discovered it could be simplified to "dns_client_mlcache#" .. tostring(opts). 

However, there's a potential issue here. In Lua, `opts` could be created anew each time by the caller,so in Lua-land, they are distinct objects each time. But we can't ensure they have different memory addresses in C-land. So, different objects might end up with the same hash value. If the previous opts table has been garbage collected and its memory space is reused by a later one.